src/plugins/dashboard/public/services/dashboard_content_management/lib/save_dashboard_state.test.ts

```diff
@@ -0,0 +1,136 @@
+  if (options.id === undefined) {
+    return { item: { id: 'newlyGeneratedId' } };
+  }
+  return { item: { id: options.id } };
+});
+
+const allServices = {
+  data,
+  embeddable,
+  notifications,
+  contentManagement,
+  initializerContext,
+  savedObjectsTagging,
+  dashboardSessionStorage,
+};
+data.query.timefilter.timefilter.getTime = jest.fn().mockReturnValue({ from: 'then', to: 'now' });
+embeddable.extract = jest
+  .fn()
+  .mockImplementation((attributes) => ({ state: attributes, references: [] }));
+
+describe('Save dashboard state', () => {
+  beforeEach(() => {
+    jest.clearAllMocks();
+  });
+
+  it('should save the dashboard using the same ID', async () => {
+    const result = await saveDashboardState({
+      currentState: {
+        ...getSampleDashboardInput(),
+        title: 'BOO',
+      } as unknown as DashboardContainerInput,
+      lastSavedId: 'Boogaloo',
+      saveOptions: {},
+      ...allServices,
+    });
+
+    expect(result.id).toBe('Boogaloo');
+    expect(allServices.contentManagement.client.create).toHaveBeenCalledWith(
+      expect.objectContaining({
+        options: expect.objectContaining({ id: 'Boogaloo', overwrite: true }),
+      })
+    );
+    expect(allServices.notifications.toasts.addSuccess).toHaveBeenCalledWith({
+      title: `Dashboard 'BOO' was saved`,
+      'data-test-subj': 'saveDashboardSuccess',
+    });
+  });
+
+  it('should save the dashboard using a new id, and return redirect required', async () => {
+    const result = await saveDashboardState({
+      currentState: {
+        ...getSampleDashboardInput(),
+        title: 'BooToo',
+      } as unknown as DashboardContainerInput,
+      lastSavedId: 'Boogaloonie',
+      saveOptions: { saveAsCopy: true },
+      ...allServices,
+    });
+
+    expect(result.id).toBe('newlyGeneratedId');
+    expect(result.redirectRequired).toBe(true);
+    expect(allServices.contentManagement.client.create).toHaveBeenCalledWith(
+      expect.objectContaining({
+        options: expect.objectContaining({ id: undefined, overwrite: true }),
+      })
+    );
+    expect(allServices.notifications.toasts.addSuccess).toHaveBeenCalledWith({
+      title: `Dashboard 'BooToo' was saved`,
+      'data-test-subj': 'saveDashboardSuccess',
+    });
+  });
+
+  it('should generate new panel IDs for dashboard panels when save as copy is true', async () => {
+    const result = await saveDashboardState({
+      currentState: {
+        ...getSampleDashboardInput(),
+        title: 'BooThree',
+        panels: { idOne: { type: 'boop' } },
+      } as unknown as DashboardContainerInput,
+      lastSavedId: 'Boogatoonie',
+      saveOptions: { saveAsCopy: true },
+      ...allServices,
+    });
+
+    expect(result.id).toBe('newlyGeneratedId');
+
+    expect(allServices.contentManagement.client.create).toHaveBeenCalledWith(
+      expect.objectContaining({
+        data: expect.objectContaining({ panelsJSON: expect.not.stringContaining('idOne') }),
```

Comment:
I'm slightly concerned about the possibility of a randomly generated id that might match these five characters causing flakiness in this test. Should we consider a longer string? e.g. `neverGonnaGetThisId`

Reply:
@nickpeihl Good point! Using `stringContaining` here makes that much more likely, as random ids contain sequences of up to 12 characters. Addressed in 44ec3258bc6